neuralmonkey/learning_utils.py

```diff
@@ -44,71 +46,85 @@ def training_loop(tf_manager: TensorFlowManager,
         epochs: Number of epochs for which the algoritm will learn.
         trainer: The trainer object containg the TensorFlow code for computing
-            the loss and optimization operation.
-        train_dataset:
-        val_dataset:
-        postprocess: Function that takes the output sentence as produced by the
-            decoder and transforms into tokenized sentence.
+                 the loss and optimization operation.
+        batch_size: how many examples are trained at once.
+        train_dataset: Dataset used for training
         log_directory: Directory where the TensordBoard log will be generated.
-            If None, nothing will be done.
+                       If None, nothing will be done.
         evaluators: List of evaluators. The last evaluator is used as the main.
-            An evaluator is a tuple of the name of the generated series, the
-            name of the dataset series the generated one is evaluated with and
-            the evaluation function. If only one series names is provided, it
-            means the generated and dataset series have the same name.
+                    An evaluator is a tuple of the name of the generated
+                    series, the name of the dataset series the generated one is
+                    evaluated with and the evaluation function. If only one
+                    series names is provided, it means the generated and
+                    dataset series have the same name.
+        runners: List of runners for logging and evaluation runs
+        val_dataset: Dataset used for evaluation of validation error
+        val_datasets: List of datasets used for evaluation of validation error.
+                      You must specify either val_dataset or val_datasets.
+        test_datasets: List of datasets used for testing
+        logging_period: after how many batches should the logging happen
+        validation_period: after how many batches should the validation happen
+        val_preview_input_series: inputs of examples to preview
+        val_preview_output_series: outputs of examples to preview
+        val_preview_num_examples: how many examples should be printed during
+                                  validation
+        train_start_offset: how many lines from the training dataset should be
+                            skipped
+        runners_batch_size: batch size of runners. It is the same as batch_size
+                            if not specified
+        initial_variables: variables used for initialization, for example for
+                           continuation of training
+        postprocess: Function that takes the output sentence as produced by the
+                     decoder and transforms into tokenized sentence.
     """
+    if val_dataset is None and val_datasets is None:
+        raise AssertionError("You must specify either val_dataset or val_datasets.")
+    if val_datasets is None:
+        val_datasets = [val_dataset]
+
```

Comment:
udělal bych z `val_dataset` deprecated config, kterej se v příští větší verzi vyhodí. Čili logiku těchto čtyř řádků přeuspořádat následovně: 

Je-li něco ve `val_dataset`, zahlásit `warn`, že je to deprecated, 
chcípnout, pokud je definovaný `val_datasets` a nastavit `val_datasets` na `[val_dataset]`. Chcípnout proto, že konfigurák, kde je obojí, je nepřehlednej a nejspíš obsahuje chybu.
Nakonec chcípnout, je-li `val_datasets` prázdný nebo `None`.

Mimochodem, jak to je napsaný teď, to případný přebytečný `val_dataset` odignoruje, což rozhodně nechcem